blockchain/treasury.go

```diff
@@ -0,0 +1,1019 @@
+		values:  make([]treasuryValue, 0, len(msgBlock.Transactions)*2),
+	}
+	trsyLog.Tracef("dbPutTreasuryBalance: %v start balance %v",
+		node.hash.String(), balance)
+	for _, v := range msgBlock.STransactions {
+		if stake.IsTAdd(v) {
+			// This is a TAdd, pull amount out of TxOut[0].  Note
+			// that TxOut[1], if it exists, contains the change
+			// output. We have to ignore change.
+			tv := treasuryValue{
+				typ:    treasuryValueTAdd,
+				amount: v.TxOut[0].Value,
+			}
+			ts.values = append(ts.values, tv)
+			trsyLog.Tracef("  dbPutTreasuryBalance: balance TADD "+
+				"%v", tv.amount)
+		} else if stake.IsTreasuryBase(v) {
+			tv := treasuryValue{
+				typ:    treasuryValueTBase,
+				amount: v.TxOut[0].Value,
+			}
+			ts.values = append(ts.values, tv)
+			trsyLog.Tracef("  dbPutTreasuryBalance: balance "+
+				"treasury base %v", tv.amount)
+		} else if stake.IsTSpend(v) {
+			// This is a TSpend, pull values out of block. Skip
+			// first TxOut since it is an OP_RETURN.
+			var totalOut int64
+			for _, vv := range v.TxOut[1:] {
+				tv := treasuryValue{
+					typ:    treasuryValueTSpend,
+					amount: -vv.Value,
+				}
+				trsyLog.Tracef("  dbPutTreasuryBalance: "+
+					"balance TSPEND %v", tv.amount)
+				ts.values = append(ts.values, tv)
+				totalOut += vv.Value
+			}
+
+			// Fees are supposed to be stored as negative amounts
+			// in a treasuryValue value, so calculate it backwards
+			// from the usual way of `in - out`.
+			fee := totalOut - v.TxIn[0].ValueIn
+			tv := treasuryValue{
+				typ:    treasuryValueFee,
+				amount: fee,
+			}
+			trsyLog.Tracef("  dbPutTreasuryBalance: "+
+				"balance fee %v", tv.amount)
+			ts.values = append(ts.values, tv)
+		}
+	}
+
+	hash := block.Hash()
+	return dbPutTreasuryBalance(dbTx, *hash, ts)
+}
+
+// dbPutTSpend inserts the TSpends that are included in this block to the
+// database.
+func (b *BlockChain) dbPutTSpend(dbTx database.Tx, block *dcrutil.Block) error {
+	hash := block.Hash()
+	msgBlock := block.MsgBlock()
+	trsyLog.Tracef("dbPutTSpend: processing block %v", hash)
+	for _, v := range msgBlock.STransactions {
+		if !stake.IsTSpend(v) {
+			continue
+		}
+
+		// Store TSpend and the block it was included in.
+		txHash := v.TxHash()
+		trsyLog.Tracef("  dbPutTSpend: tspend %v", txHash)
+		err := dbUpdateTSpend(dbTx, txHash, *hash)
+		if err != nil {
+			return err
+		}
+	}
+
+	return nil
+}
+
+// TreasuryBalanceInfo models information about the treasury balance as of a
+// given block.
+type TreasuryBalanceInfo struct {
+	// BlockHeight is the height of the requested block.
+	BlockHeight int64
+
+	// Balance is the balance of the treasury as of the requested block.
+	Balance uint64
+
+	// Updates specifies all additions to and spend from the treasury in the
+	// requested block.
+	Updates []int64
+}
+
+// TreasuryBalance returns treasury balance information as of the given block.
+func (b *BlockChain) TreasuryBalance(hash *chainhash.Hash) (TreasuryBalanceInfo, error) {
```

Comment:
Dunno. Did fix it.

Reply:
Doesn't matter too much yet without profiling results to back it up, but realistically speaking, it would be faster to avoid the alloc here because the struct is only `8+8+24=40` bytes and that is **significantly** faster to copy versus all the GC overhead.  In other words, it was better without being a pointer.